### PR TITLE
Fix #895: Add Dotty implicit functions to the parser

### LINKED ISSUE #895: Add Dotty implicit function to the parser
https://scastie.scala-lang.org/XxaGpCRKRJqOpz2SZo0R5Q

work without instrumentation:

```scala
object B {
	trait A
	type X = implicit A => Int

  def main(arg: Array[String]): Unit = {
		println("hello")
  }
}
```

### PR DESCRIPTION
**Pull request by jonas**

### CODE REVIEW

File: scalameta/dialects/shared/src/main/scala/scala/meta/dialects/package.scala [hunk 1]
```diff
@@ -23,6 +23,9 @@ import scala.compat.Platform.EOL
   // Are extractor varargs specified using colons, i.e. is `case Extractor(xs: _*)` legal or not?
   allowColonForExtractorVarargs: Boolean,
 
+  // Are implicit functions supported by this dialect?
+  allowImplicitFunctions: Boolean,
```
Comment (xeno-by) [inline scalameta/dialects/shared/src/main/scala/scala/meta/dialects/package.scala]: I suggest we call this `allowImplicitFunctionTypes`.

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 2]
```diff
@@ -3138,7 +3138,14 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
...
     token match {
-      case Equals() => next(); aliasType()
+      case Equals() =>
+        next()
+        if (token.is[KwImplicit] && dialect.allowImplicitFunctions) {
+          next()
+          Defn.Type(mods ++ immutable.Seq(Mod.Implicit()), name, tparams, typ())
+        } else {
+          aliasType()
+        }
```
Comment (xeno-by) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: Are implicit function types supported only in type aliases?

Comment (jonas) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: Looking at [the tests](https://github.com/lampepfl/dotty/pull/1775/files#diff-21900f439c1e991ad21e8c9be8c80fe8R6) in the PR linked from https://www.scala-lang.org/blog/2016/12/07/implicit-function-types.html there's also 

```
val i1 = (implicit (s: String) => s.length > 2)
val x: implicit String => Boolean = { implicit (s: String) => s.length > 2 }
```

**Comment by olafurpg** [comment]
The CI failures are unrelated to this PR. It seems some http requests to Github failed. I'm gonna restart the job to try and get that green icon :)

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 3]
```diff
@@ -988,15 +988,29 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
...
+            Type.ImplicitFunction(params, body)
+          case t =>
+            syntaxError(s"$dialect does not support implicit function types", at = t)
+        }
+      } else {
+        typRest()
+      }
+    }
+
+    def typRest(): Type = autoPos {
```
Comment (olafurpg) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: I don't know if `typRest` is a good name or not, wdyt @xeno-by ?

**Comment by allanrenucci** [comment]
Here are some test cases from the dotty test suite:
```scala
def foo0(): implicit Int => Int = ???
val foo01: implicit (String, Int) => Int = ???
def foo[A, B]: implicit A => implicit B => (A, B) = ???

def foo1(f: Int => implicit Int): Int = ??? // error
def foo2(): implicit Int = ??? // error
def foo3(): Int => implicit Int = ??? // error
```

**Comment by b-studios** [comment]
That's great, thanks @jonas!

**Comment by jonas** [comment]
@allanrenucci Great, added additional tests.

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 4]
```diff
@@ -988,15 +988,29 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
...
      *  ExistentialDcl    ::= type TypeDcl | val ValDcl
      *  }}}
      */
-    def typ(): Type = autoPos {
+    def typ(): Type = {
+      if (token.is[KwImplicit] && dialect.allowImplicitFunctionTypes) {
+        next()
+        typRest() match {
+          case Type.Function(params, body) =>
+            Type.ImplicitFunction(params, body)
```
Comment (xeno-by) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: What about the position for this tree?

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala (hunk 3 above)
Comment (xeno-by) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: Sounds allright.

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 5]
```diff
@@ -988,15 +988,29 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
...
+     *  Type ::= [`implicit'] InfixType `=>' Type
+     *         | [`implicit'] `(’ [`=>’ Type] `)’ `=>' Type
      *         | `[' [`=>' Type] `]' `=>' Type
      *         | InfixType [ExistentialClause]
      *  ExistentialClause ::= forSome `{' ExistentialDcl {semi ExistentialDcl}} `}'
      *  ExistentialDcl    ::= type TypeDcl | val ValDcl
      *  }}}
      */
-    def typ(): Type = autoPos {
+    def typ(): Type = {
```
Comment (olafurpg) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: Can we add `autoPos` here?

Comment (jonas) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: I've restored `autoPos` on `typ()`

**Comment by xeno-by** [comment]
LGTM

**Pull request merged**